### PR TITLE
Use dynamic library loading for detecting the correct path for links

### LINKED ISSUE #2719: [windows] When creating explorer favorite use more specific windows functions
I'm experiencing problems with the creation of the favourite on Windows 7, using folder redirection. I made some research for the problem and found out that client tries to create a link on the wrong place. My "Links" directory is on a cifs share: \\server\homes\username\Links. This wrong behavior is caused by the function setupFavLink_private in [utility_win.cpp](https://github.com/owncloud/client/blob/master/src/libsync/utility_win.cpp)

At the moment in function _setupFavLink_private_ at  [line 24](https://github.com/owncloud/client/blob/master/src/libsync/utility_win.cpp#L24) the function

```
    SHGetSpecialFolderPath(0, path, CSIDL_PROFILE, FALSE);
```

is used.
[MSDN documentation](http://msdn.microsoft.com/en-us/library/windows/desktop/bb762204%28v=vs.85%29.aspx) says that [ShGetFolderPath](http://msdn.microsoft.com/en-us/library/windows/desktop/bb762181%28v=vs.85%29.aspx) should be used. But says also that ShGetFolderPath is deprecated. A note states:

> Note  As of Windows Vista, this function is merely a wrapper for SHGetKnownFolderPath. The CSIDL value is translated to its associated KNOWNFOLDERID and then SHGetKnownFolderPath is called. 
> New applications should use the known folder system rather than the older CSIDL system, which is supported only for backward compatibility.

Please note that SHGetKnownFolder is a function available only on Vista and later. But, you can also note that also favourites on explorer are available since that windows release.

I made a test and tried to print the path value that is returned by the function call:

```
    SHGetKnownFolderPath(FOLDERID_Links, 0, NULL, path);
```

and i get (correctly) the value of _\\server\homes\username\Links_.
I printed also the output of 

```
    SHGetSpecialFolderPath(0, path, CSIDL_PROFILE, FALSE);
```

and i got _C:\Users\username\_, that is correctly my profile home, but the path where the function will create the link is _C:\Users\username\Links_, that is wrong.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/7876774-windows-when-creating-explorer-favorite-use-more-specific-windows-functions?utm_campaign=plugin&utm_content=tracker%2F216457&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F216457&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by remixtj**
Fixes #2719

### CODE REVIEW

File: src/libsync/utility_win.cpp [hunk 1]
```diff
@@ -14,19 +14,60 @@
...
 
 static const char runPathC[] = "HKEY_CURRENT_USER\\Software\\Microsoft\\Windows\\CurrentVersion\\Run";
 
+typedef HRESULT (WINAPI *SHGetKnownFolderPathFun)(
+    const GUID &rfid, 
+    DWORD dwFlags, 
+    HANDLE hToken, 
+    PWSTR *ppszPath
+);
+typedef HRESULT (WINAPI *SHGetFolderPathFun)(
```
Comment (jturcotte) [inline src/libsync/utility_win.cpp]: Thanks for the update. You won't need that part anymore either I think though.

**Comment by jturcotte** [comment]
Did you try the patch on XP yourself? If not I'll have to spend some time to try it before we can merge it.

**Comment by remixtj** [comment]
I'm now preparing an XP machine to test if works. I'll update you when done

**Comment by remixtj** [comment]
Did a test, the setup worked correctly, no errors reported. Client working without evident issues on windows xp.

**Comment by jturcotte** [comment]
If it called setupFavLink_private and it didn't crash that should already be good. Thanks for testing : )

**Comment by jturcotte** [comment]
Hi, so I just realised that you guys need to sign the CLA before we can accept your patches.

The info is there: https://owncloud.org/contribute/agreement/
If you can't sign it we'll unfortunately have to revert the merge request.

**Comment by jturcotte** [comment]
Hi @remixtj, any luck with the CLA?

**Comment by remixtj** [comment]
Waiting for my boss. I hope to have good news in a week. Informally told me
that there are no problems signing. The real problem is finding which is
the most appropriate person that has to sign.
Il giorno 15/giu/2015 16:13, "Jocelyn Turcotte" <notifications@github.com>
ha scritto:

**Comment by guruz** [comment]
@remixtj Any news? :)

**Comment by jturcotte** [comment]
Hi @remixtj, sadly if we can't get that CLA signed before the beta we'll have to revert this merge request.This should happen in little more than a week.

Sorry, it was my mistake to accept it before this was cleared. If you manage to sign it afterward we also can re-merge it for the next release.

**Comment by remixtj** [comment]
I'm sorry, my manager  forgotto tosend me the document signed. I'out of
office for 15 days, i'll try to provide you asap.

I'll make a new pull request when available. Thank you

On Monday, July 20, 2015, Jocelyn Turcotte <notifications@github.com> wrote:
> Hi @remixtj, sadly if we can't get that CLA signed before the beta we'll
have to revert this merge request.This should happen in little more than a
week.
>
> Sorry, it was my mistake to accept it before this was cleared. If you
manage to sign it afterward we also can re-merge it for the next release.
>
> —
> Reply to this email directly or view it on GitHub.<
https://ci5.googleusercontent.com/proxy/4Ut0VyMwWThO8bEVzZfJtcw4REhtbmX0hbmsiEDXt24WzXXoDOdz-wZe8A9fRb3ER36_Ls2qLBazn_79RSNeS7eQdupVhKcv45-Q6DW1ICvp063HTxU2TBoVXYC-TOPl8TJGQvk-uSEtOxjvIMhMNcOsEDNWKw=s0-d-e1-ft#https://github.com/notifications/beacon/ABQEgW2F_91oiwZcP22kFzncL2bvo-Nnks5ofJ4zgaJpZM4Dpe88.gif
>

-- 
"E' assurdo impiegare gli uomini di intelligenza eccellente per fare
calcoli che potrebbero essere affidati a chiunque se si usassero delle
macchine"
Gottfried Wilhelm von Leibnitz, Filosofo e Matematico (1646-1716)

"Internet è la più grande biblioteca del mondo.
Ma il problema è che i libri sono tutti sparsi sul pavimento"
John Allen Paulos, Matematico (1945-vivente)

Luca 'remix_tj' Lorenzetto, http://www.remixtj.net , <
lorenzetto.luca@gmail.com>

**Comment by guruz** [comment]
PR reverted for now..

**Comment by guruz** [comment]
Moved to 2.0.1

**Pull request merged**